Click-to-Pregrasp

hello-amal opened the pull request:
# Description

This PR implements a feature where the user is able to click a point on the Realsense stream, and if the point is within the robot's reach, it will align with the point with either a horizontal or vertical gripper (user-specified). The aim of this feature is to prevent the iterative and time-consuming movements required for an operator to properly align the gripper with an object to grasp it.

# Upgrade Info

See README.md for information on installing the additional requirements and generating the specialized URDF.

# Testing procedure

- [x] Stretch RE3 with Dex Wrist and D405:
    - [x] Launch the interface, add a Realsense Camera with buttons displayed below.
    - [x] Click on an object on the floor that is within the graspable region with a horizontal grasp, verify it works.
    - [x] Click on an object on the floor that is within the graspable region with a vertical grasp, verify it works.
    - [x] Click on an object on a table that is within the graspable region with a horizontal grasp, verify it works.
    - [x] Click on an object on a table that is within the graspable region with a vertical grasp, verify it works.
    - [x] Click on an object outside the graspable region with a horizontal grasp, verify it fails.
    - [x] Click on an object outside the graspable region with a vertical grasp, verify it fails.
- [x] Stretch3 with tablet:
    - [x]  Verify that the click-to-pregrasp buttons don't appear.
- [x] Stretch RE2 with Dex Wrist and D405:
    - [x] Repeat tests as above.
- [x] Stretch RE2 with Dex Wrist and Beta Teleop Kit:
    - [x] Repeat tests as above.
- [x] Stretch RE2 without Dex Wrist:
    - [x] Verify that the "Select Object" and "Align Gripper with Object" buttons do not appear.

# Before opening a pull request

From the top-level of this repository, run:

- [x] `pre-commit run --all-files`

# To merge

- [x] `Squash & Merge`

hello-amal commented:
@hello-cpaxton This is ready for a review of the functionality (`move_to_pregrasp.py`, `stretch_ik_control.py`). I'll be working on the UX this week.

hello-cpaxton commented on nodes/move_to_pregrasp.py:
```diff
@@ -0,0 +1,952 @@
+    TERMINAL = 8
+
+    @staticmethod
+    def get_state_machine(
```
Do we have a standard way to set up state machines like this? I dont love using enums

hello-cpaxton replied:
Not something you should change or anything

hello-amal replied:
Don't have a standard way of doing it. I liked your `task.py` formulation, and if we want to move a bit more generic, in the past I've really liked behavior trees.

For this node in particular, I've considered making a separate "state" class, and each of these states is an instance of the class, and each state owns the logic of creating its own motion generator (as opposed to it being done in `execute_action` below). However, I feel that's low-priority at the moment.

hello-cpaxton replied:
Yeah I also like behavior trees

hello-cpaxton replied:
Agree

hello-cpaxton commented on nodes/move_to_pregrasp.py:
```diff
@@ -0,0 +1,952 @@
+        ]
+
+
+class MoveToPregraspNode(Node):
```
Do you see any performance hit from using so many ROS nodes?

hello-amal replied:
This feature adds one node (`/move_to_pregrasp`). The complete list of running nodes is:
```
/amcl
/behavior_server
/bt_navigator
/bt_navigator_navigate_through_poses_rclcpp_node
/bt_navigator_navigate_to_pose_rclcpp_node
/camera
/configure_video_streams
/controller_server
/d435i_accel_correction_node
/global_costmap/global_costmap
/gripper_camera
/joint_state_publisher
/joint_trajectory_action
/laser_filter
/laser_filter
/laser_filter
/lifecycle_manager_localization
/lifecycle_manager_navigation
/local_costmap/local_costmap
/map_server
/move_to_pregrasp
/navigation_camera
/planner_server
/robot_state_publisher
/rosapi
/rosapi_params
/rosbridge_websocket
/sllidar_node
/static_transform_publisher_3cyXiogXh3YjhR0C
/stretch_driver
/tf2_web_republisher
/transform_listener_impl_594129e43ba0
/transform_listener_impl_5d048bf95800
/transform_listener_impl_5fe79ce6f040
/transform_listener_impl_61f99bae9710
/transform_listener_impl_63b81553c220
/waypoint_follower
```

I haven't found any performance hit with the "low resolution" D435i. I did find a performance hit when I increased resolution of the D435i, presumably because both `/configure_video_streams` and `/move_to_pregrasp` subscribe to its streams. To address that, I did try to combine those two nodes into one, but that resulted in a priority inversion issue where action calls were not received for very long, presumably because camera messages were always ready to pop off of the waitlist before the action invokation.

Since we're switching to a python-backend web interface, I don't think it's worthwhile overfitting to ROS issues (as long as it doesn't hard functionality/usability/reliability).

hello-cpaxton commented on nodes/move_to_pregrasp.py:
```diff
@@ -0,0 +1,952 @@
+    the object the pixel is a part of.
+    """
+
+    DISTANCE_TO_OBJECT = 0.1  # meters
```
this is the distance to move from clicked pixel? clearer comment maybe?

hello-cpaxton commented on nodes/move_to_pregrasp.py:
```diff
@@ -0,0 +1,952 @@
+
+        # Create the inverse jacobian controller to execute motions
+        # TODO: Figure out where the specialized URDF should go!
+        urdf_abs_path = "/home/hello-robot/stretchpy/src/stretch/motion/stretch_base_rotation_ik.urdf"
```
hard coded paths are bad - maybe we add a utility to stretchpy to find this?

hello-cpaxton commented on nodes/move_to_pregrasp.py:
```diff
@@ -0,0 +1,952 @@
+            f"Clicked pixel after inverse transform (camera frame): {(u, v)}"
+        )
+
+        # Get the latest Realsense messages
```
this single huge function seems like quite a lot

hello-cpaxton commented on nodes/pinocchio_ik_solver.py:
```diff
@@ -0,0 +1,203 @@
+        desired_ee_pose = pinocchio.SE3(
+            R.from_quat(quat_desired).as_matrix(), pos_desired
+        )
+        while True:
```
rclpy.ok()

hello-cpaxton replied:
oops wrong class

hello-cpaxton commented on nodes/move_to_pregrasp.py:
```diff
@@ -0,0 +1,953 @@
+        goal_pose.pose.orientation = Quaternion(x=0.0, y=0.0, z=0.0, w=1.0)
+
+        # Convert to base frame
+        try:
```
can clean up this class by moving all of this into a helper function

hello-cpaxton commented on nodes/move_to_pregrasp.py:
```diff
@@ -0,0 +1,953 @@
+        if not horizontal_grasp:
+            # For vertical grasp, the goal orientation is also rotated +90deg around y
+            rot = quaternion_multiply(rot, quaternion_about_axis(np.pi / 2, [0, 1, 0]))
+        goal_pose_base.pose.orientation = Quaternion(
```
we can probably tweak the formatting rules so this isnt 3 lines

hello-cpaxton commented on nodes/move_to_pregrasp.py:
```diff
@@ -0,0 +1,953 @@
+                    # Only care about x and yaw in base frame. y and z are adjusted by
+                    # arm length/lift
+                    cartesian_mask = np.array([True, False, False, False, False, True])
+                if MoveToPregraspState.HEAD_PAN in concurrent_states:
```
why is current states multiple things? i guess you can execute multiple steps at once in this model?

hello-amal replied:
Yes, this is specifically because of a feature Blaine requested (that is part of this PR) to rotate the head pan to counteract base rotation. So those states are executed at the same time.

More generally, if in the future we want to make this less step-wise and move multiple joints together, we can use this "combining states" feature.

hello-cpaxton commented on nodes/stretch_ik_control.py:
```diff
@@ -0,0 +1,1217 @@
+"""
+This module contains the StretchIKControl class, which allows users to compute Stretch's FK and IK,
```
This would be nice to bring out into a more general library like stretchpy

hello-cpaxton commented on nodes/stretch_ik_control.py:
```diff
@@ -0,0 +1,1217 @@
+        self.controllable_joints = [
+            Joint.BASE_ROTATION,
+            Joint.ARM_LIFT,
+            Joint.ARM_L0,  # The URDF uses the most distal of the telescoping joints
```
cannot do wrist?

hello-amal replied:
Not via inverse jacobian velocity control. `controllable_joints` are the joints that can be moved by inverse jacobian velocity control. I'll clarify that in a comment.

In practice, because the wrist also impacts the same end effector pose dimensions as the above three, we were getting some Jacobians that were attempting to use them to compensate for each other, e.g., trying to pitch up the wrist if arm lift overshot.

hello-cpaxton commented on nodes/stretch_ik_control.py:
```diff
@@ -0,0 +1,1217 @@
+        self.joint_vel_abs_lim: Dict[Joint, Tuple[float, float]] = {}
+
+        # Get the velocity limits for the controllable joints
+        for joint_name in Joint:
```
might be nice to break this up into separate functions or something for readability

hello-cpaxton commented on nodes/stretch_ik_control.py:
```diff
@@ -0,0 +1,1217 @@
+                min_abs_vel = robot_params["lift"]["motion"][speed_profile_slow_str][
+                    "vel_m"
+                ]
+            elif joint_name == Joint.ARM_L0 or joint_name == Joint.COMBINED_ARM:
```
Should we catch an error to make sure both of these dont happen

hello-amal replied:
There's no harm in both of these happening. In fact, it ends up being helpful to have both joints in the velocity limits dictionary, because the URDF uses `joint_arm_l0` but we command `joint_arm`, so it gives us more flexibility to apply velocity limits at both layers.

hello-cpaxton commented on nodes/stretch_ik_control.py:
```diff
@@ -0,0 +1,1217 @@
+                    speed_profile_slow_str
+                ]["vel"]
+            elif joint_name == Joint.WRIST_ROLL:
+                max_abs_vel = robot_params["wrist_roll"]["motion"][speed_profile_str][
```
it feels like there must be a cleaner way to write this loop

hello-cpaxton replied:
like a dictionary from joint.wrist_roll to "wrist_roll"

hello-cpaxton commented on nodes/stretch_ik_control.py:
```diff
@@ -0,0 +1,1217 @@
+        yield MotionGeneratorRetval.FAILURE if check_cancel() else MotionGeneratorRetval.SUCCESS
+        return
+
+    def set_navigation_mode(self, timeout: Duration, rate_hz: float = 10.0) -> bool:
```
this logic feels like it's going to appear in a lot of applications, would be nice to factor out

hello-amal replied:
I kept it in for now, but am happy to take lead on factoring it out when we know where we want to put it or have other applications to use it.

hello-cpaxton commented on nodes/stretch_ik_control.py:
```diff
@@ -0,0 +1,1217 @@
+        if isinstance(target_frame, Frame):
+            target_frame = target_frame.value
+        try:
+            pose_transformed = self.tf_buffer.transform(
```
might be nice to have tf wrapper for getting poses

hello-cpaxton reviewed (commented):
looks good, mostly reviewed stretch_ik_control and move_to_pregrasp

hello-binit commented:
@hello-amal jFYI, I tried running this branch and saw the following errors. You may need to add the `constants` and `helpers` python packages as ROS deps in package.xml. Or they can be [added here](https://github.com/hello-robot/stretch_install/blob/2e38dd24e779bdae7bad176adec49c1e3068b091/factory/22.04/stretch_create_ament_workspace.sh#L68-L69).

![Screenshot from 2024-06-12 23-01-35](https://github.com/hello-robot/stretch_web_teleop/assets/64861565/24b8409d-4494-4e78-9a80-60dcf05e6ce2)

hello-amal commented:
Thanks for surfacing that @hello-binit . I was able to recreate it when rebuilding from scratch, and the latest commit should address it -- try it out when you get a chance.

hello-amal commented:
Thanks for the comments @hello-cpaxton ! I addressed them and cleaned up th code. @hello-vinitha , you can review it now when you get a chance. This has been thoroughly tested on the Stretch 3 (with the dex wrist and with the tablet), not yet on earlier configurations.

hello-vinitha commented:
@hello-amal I tested this PR on a Stretch 2 with the upgraded teleop kit and dex wrist. The arm collides with the base when the (horizontal) pre-grasp z position is below the base.

hello-vinitha reviewed (approved):
Tested on the various Stretch 2 configurations. All tests pass!
